[NNC] refactor cuda half support to more general file

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #47374 [NNC] Enable unary op cpu testing
* **#47373 [NNC] refactor cuda half support to more general file**
* #47372 [NNC] More cpu tests
* #47371 [NNC] Add more CPU Tests
* #47370 [NNC] Fix llvm min lowering for int inputs
* #47369 Add more CPU tests
* #46951 [TensorExpr][CPU] Fix bool -> int casting
* #46950 Small changes/cleanup
* #46949 Force LLVM Compilation for CPU Tests

Differential Revision: [D24805246](https://our.internmc.facebook.com/intern/diff/D24805246)

Commit:
refactor cuda half support to more general file

[ghstack-poisoned]

Commit:
Update on "refactor cuda half support to more general file"




[ghstack-poisoned]

Commit:
Update on "refactor cuda half support to more general file"




[ghstack-poisoned]

Commit:
Update on "refactor cuda half support to more general file"




[ghstack-poisoned]

Commit:
Update on "[NNC] refactor cuda half support to more general file"




[ghstack-poisoned]

bertmaher commented on torch/csrc/jit/tensorexpr/half_support.h:
```diff
@@ -0,0 +1,71 @@
+namespace jit {
+namespace tensorexpr {
+
+// Walk the Statment looking for Half size loads/stores.
```
Probably best to do in a follow-up diff but it bugs me that something called "Checker" actually does mutation, and the doc comment doesn't give any indication of what it's doing either.  It'd be nice to rename and add a clearer comment.

eellison replied:
This comment (and your other comments) will be are addressed when I rebase onto  by Nick's PR https://github.com/pytorch/pytorch/pull/47448

nickgg replied:
Yeah this is fixed already :) (you're right though)

bertmaher commented on torch/csrc/jit/tensorexpr/half_support.h:
```diff
@@ -0,0 +1,71 @@
+    const Expr* new_val = v->value()->accept_mutator(this);
+
+    if (v->value()->dtype().scalar_type() == ScalarType::Half) {
+      // TODO discards lanes.
```
Hmmm is this `discard lanes` comment still true, or is it stale?  The line below seems to preserve lanes, but...?

nickgg replied:
Should be fixed in master.

bertmaher reviewed (approved):
Cool with landing this as-is since it's code motion, just a couple notes on this file in general.

Commit:
Update on "[NNC] refactor cuda half support to more general file"




[ghstack-poisoned]

Commit:
Update on "[NNC] refactor cuda half support to more general file"




[ghstack-poisoned]

nickgg commented on torch/csrc/jit/tensorexpr/half_support.h:
```diff
@@ -1,14 +1,15 @@
 #pragma once
 
-#include <torch/csrc/jit/codegen/fuser/cuda/resource_strings.h>
```
Cuda codegen needs this. You'll have to add it back there.

eellison replied:
This breaks the build when you're not using cuda... Why does it need to be in this file?

nickgg reviewed (dismissed):
Let's make sure we don't break the Cuda backend since that's the bit thats turned on.

eellison commented:
> Let's make sure we don't break the Cuda backend since that's the bit thats turned on.

Sure, Ci has to be green before i'm going to land

Commit:
Update on "[NNC] refactor cuda half support to more general file"


Differential Revision: [D24805246](https://our.internmc.facebook.com/intern/diff/D24805246)

[ghstack-poisoned]

Commit:
Update on "[NNC] refactor cuda half support to more general file"


Differential Revision: [D24805246](https://our.internmc.facebook.com/intern/diff/D24805246)

[ghstack-poisoned]

nickgg reviewed (approved):
🚢
